anymaniax/orval
# WIP: convert MSW to use a syntax tree and use it to sort keys

## Status

**WIP**. This is phase 1 exploring the general direction. I still need to fix an outstanding issue, and update the tests.

## Description

This PR uses issue https://github.com/anymaniax/orval/issues/703 as a test bed for adding an AST-based intermediate layer. It does a basic switch-over where it makes sense, and collects formatting logic into one place. Right now, a lot of fields are simple 'primitive' indicating verbatim code to be inserted into the file. We also support oneOf, anyOf, allOf, and object (the basis for which this PR is based on).

This also opens us up for a 'function' syntax type down the line which we can use to extract certain schemas into reusable functions which will help us speed up our code as well as improve the code reuse in the output making it less verbose and more readable. That part will come in a later PR.

_Not_ covered in this PR is the question of global code reuse. I assume the syntax tree intermediate AST will probably have some parts that are useful to the rest of the code, but I don't know yet if it is enough to build it with that in mind.

## Extra comments

- I have reworked the precommit hook to use turbo for parallelism and to add typechecking. The hook runs faster and does more 😄 
- I am planning on adding a test harness for msw (probably snapshots)
- Is the code output considered a 'breaking' change? Reordering the fields by default will cause noise.
- It would be nice to include the samples packages in the turbo workspace but I understand why that is not the case (lots and lots and lots of dependencies for one)
- It would also be nice to include the test packages

## Todos


## Steps to Test or Reproduce

Outline the steps to test or reproduce the PR here.

1. run turbo dev
2. look at the outputed msw files and see they are ordered correctly
